### ISSUE TITLE
ETL scenesfrancophones.ca

### ISSUE THREAD

**Issue by saumier** [culturecreates/artsdata-orion#7]
Example event with JSON-LD 
https://scenesfrancophones.ca/spectacles/luc-leblanc-2533

**Comment by saumier** [comment]
@dev-aravind Hi. The first challenge is to get the list of webpages.  I understood that you are looking at Python. 

The second challenge is to get the structured data from the webpage (sometimes called 'linked data' or "RDF'). This second part should use a library that already knows how to extract structured data **because** there are too many different considerations for us to code ourselves. I cannot stress this enough.

Some of the considerations are messy HTML with invalid code and missing tags.  Another consideration is that the JSON-LD on the webpage may be in several different <script> nodes. More importantly, linked data can also be in **RDFa** or **Microdata** formats which are also very common on webpages. Finally, the server may use the headers of the request to do content negotiation, which is a more advanced method to get RDF.  Since we are looking for a solution that we can use on **multiple websites** (not just scenesfrancophones.com) it is important to not write specific code for a single website and a single serialization format.  The call to load the structured data from a url should abstract the format of the linked data.  I don't know about Python, but in Ruby this can be done with the following single line of code `RDF::Graph.load("url of webpage with structured data")`.  For example:  `RDF::Graph.load('https://scenesfrancophones.ca/diffuseurs/acfa-regionale-de-grande-prairie')`. This code handles many different possibilities of how the webpage can store the structured data, in any of the multitude of RDF serialization formats including JSON-LD, Microdata and RDFa, and also does content negotiation if the server supports it. 

Here is the complete Ruby code to load the entire website structured data into a graph given an array of urls of webpages:

```
require 'linkeddata'
graph = RDF::Graph.new
urls.each do |url|
     graph << RDF::Graph.load(url)
end
```


Once all the entities are loaded from the webpages, the output can be generated in JSON-LD as follows:
`graph.dump(:jsonld) `

The output can also be `graph.dump(:turtle)` or other serializations of RDF, and it can be queried with basic graph patterns or SPARQL to take advantage of all the RDF concepts.

In conclusion, please continue to look into Python, or any other language that you think will suit our needs, and see if there is a library to abstract structured data from the webpage that abstracts the mechanism, so that when you call the webpage url, you load all the structured data available. If there are libraries other than the Ruby ones, then I am very interested to try them out. But if not, lets pursue with Ruby ;-)

**Comment by saumier** [comment]
@dev-aravind I have merged the feature branch. You can continue in the 'main' branch until we have a functional system. 

Please try to get the workflow running such that the result of the scrape is committed to the repo.  I tried a manual run this morning but there is a Ruby version problem. Please create a .ruby-version file in the repo (which gets used by the workflow), and add a Gemfile to the repo. Finally, once you have rvm installed please create a .rvmrc file in the repo with the correct version of Ruby and a gemset. You can call the gemset the same as the repo name 'artsdata-planet-scenesfrancophones'.  From memory, the .rvmrc file should contain only the following line `rvm use 3.1.2@artsdata-planet-scenesfrancophones --create` and the .ruby-version should contain only `3.1`. If you are using a different version of Ruby please make sure the versions match everywhere.

**Comment by saumier** [comment]
@dev-aravind To test that rvm is working, when you cd into the directory, the gemset and ruby version should be set automatically. So if you do `rvm use x.x` and then cd into your directory the gemset and ruby version should change to the one set in the .rvmrc file.

**Comment by saumier** [comment]
@dev-aravind I see that you are using Thread. I would recommend removing this for 2 reasons:  1. We need to play nice with the websites and not load them with a sudden burst of traffic, 2. I would like to add tests before adding complexity, so having the simplest code is better at this stage, rather than a performant code. With the small sites we are crawling at the moment, there is no need to performance, so simplicity, unit tests, and good code practice should prioritized.

**Comment by dev-aravind** [comment]
@saumier I've removed the thread operation and used rvm for version control.
